File: model/player.ts
import {DecisionTree} from './decisionTree'
import {Decision} from './decision'
import {Board} from './board'
import {RowColumnSet} from './rowColumnSet'
import {IterationExcess} from './iterationExcess'
import {Contraddiction} from './contraddiction'
import {Inconsistency} from './inconsistency'

export class Player {
    decisionTree: DecisionTree;
    
    maxIterations = 10000;
    
    public solve(_board: Board, _maxIterations?: number, _resetDecisionTree?: boolean)
    {
        if (_maxIterations == null) {
            _maxIterations = this.maxIterations;
        }
        if (_resetDecisionTree == null) {
            _resetDecisionTree = true;
        }
        let iterations = 0;
        
        this.checkConsistency(_board);

        this.setIdentifiedValues(_board);
        if (_board.isFinished())
            return iterations;

        if (_resetDecisionTree || this.decisionTree == null)
            this.decisionTree = new DecisionTree(_board);
        
        iterations = this.iterateSolution(_board, _maxIterations);

        this.checkConsistency(_board);

        return iterations;
    }

    public tryAnotherSolution(_board: Board, _maxIterations?: number)
    {
        if (_maxIterations == null) {
            _maxIterations = this.maxIterations;
        }
        if (this.decisionTree != null)
        {
            if (!this.decisionTree.isForBoard(_board))
                throw new Error("Decision tree for a different board");
            else
                this.decisionTree.changeDecision();
        }
        return this.iterateSolution(_board, _maxIterations);
    }
        
    public bringBackToOriginalBoardLayout(_board: Board)
    {
        if (this.decisionTree != null)
        {
            if (!this.decisionTree.isForBoard(_board))
                throw new Error("Decision tree for a different board");
            this.decisionTree.reset();
        }
    }

    public undoLastDecision(_board: Board)
    {
        if (this.decisionTree != null)
        {
            if (!this.decisionTree.isForBoard(_board))
                throw new Error("Decision tree for a different board");
            this.decisionTree.undoLastDecision();
        }
    }

    public iterateSolution(_board: Board, _maxIterations: number)
    {
        let iterations = 0;
        if (_maxIterations == null) {
            _maxIterations = this.maxIterations;
        }
        while (_board.notFinished())
        {
            if (iterations > _maxIterations)
            {
                throw new IterationExcess("Max number of iterations exceeded (" + iterations + ")");
            }
            iterations++;
            this.performNextDecision(_board);
        }
        return iterations;
    }

    public performNextDecision(_board: Board)
    {
        let decision = this.decisionTree.nextDecision();
        let hasContraddiction = false;
        do
        {
            try
            {
                hasContraddiction = false;
                this.setIdentifiedValues(_board, decision);
            }
            catch (ex)
            {
                if (ex instanceof Contraddiction) {
                    console.log(ex.message);
                    decision = this.decisionTree.changeDecision();
                    hasContraddiction = true;                    
                } else {
                    throw(ex);
                }
            }
        }
        while (hasContraddiction);
    }

    public setIdentifiedValues(_board: Board, _decision?: Decision)
    {
        let performLoop;
        do 
        {
            performLoop = false;
            for (var i = 0; i < _board.cells.length; i++) {
                for(var j = 0; j < _board.cells.length; j++) {
                    let cell = _board.cells[i][j];
                    performLoop = performLoop || cell.identifyAndSetValue(_decision);
                }                
            }
        }
        while(performLoop);
    }

    public checkConsistency(_board: Board)
    {
        // first of all clean previous states
        for (var i = 0; i < _board.rows.length; i++)
        {
            _board.rows[i].isInconsistent = false;
        }
        for (var i = 0; i < _board.columns.length; i++)
        {
            _board.columns[i].isInconsistent = false;
        }
        for (var i = 0; i < _board.squareSets.length; i++)
        {
            _board.squareSets[i].isInconsistent = false;
        }
        // then check the consistency
        for (var i = 0; i < _board.rows.length; i++)
        {
            let row = _board.rows[i];
            this.checkConsistencyInRowColumnSet(row, "row");
        }
        for (var i = 0; i < _board.columns.length; i++)
        {
            let column = _board.columns[i];
            this.checkConsistencyInRowColumnSet(column, "column");
        }
        for (var i = 0; i < _board.squareSets.length; i++)
        {
            let squareSet = _board.squareSets[i];
            this.checkConsistencyInRowColumnSet(squareSet, "squareSet");
        }
    }
    
    protected checkConsistencyInRowColumnSet(_rowColumnSet: RowColumnSet, _type: string)
    {
        let oneToNine = new Array<string>(9);
        for (var i = 0; i < _rowColumnSet.cells.length; i++)
        {
            let cell = _rowColumnSet.cells[i];
            if (cell.isValueSet())
            {
                let val = cell.val;
                if (oneToNine[val-1] != null) {
                    _rowColumnSet.isInconsistent = true;
                    throw new Inconsistency("Inconsistency in " + _type + " " + _rowColumnSet.index);
                }
                else
                    oneToNine[val-1] = val.toString();
            }
        }
    }
        
}

File: SudokuClient/sudoku/Player.cs
using System;

namespace Sudoku_Core
{
	/// <summary>
	/// Summary description for Player.
	/// </summary>
	public class Player
	{
		protected DecisionTree decisionTree;

		protected static int MaxIterations = 10000;

		public Player()
		{
			//
			// TODO: Add constructor logic here
			//
		}

		public int Solve(Board _board)
		{
			return Solve(_board, MaxIterations);
		}

		public int Solve(Board _board, int _maxIterations)
		{
			return Solve(_board, _maxIterations, true);
		}

		public int TryAnotherSolution(Board _board)
		{
			return TryAnotherSolution(_board, MaxIterations);
		}

		public int TryAnotherSolution(Board _board, int _maxIterations)
		{
			if (decisionTree != null)
			{
				if (!decisionTree.IsForBoard(_board))
					throw new Exception("Decision tree for a different board");
				else
					decisionTree.ChangeDecision();
			}
			return IterateSolution(_board, _maxIterations);
		}

		public void BringBackToOriginalBoardLayout(Board _board)
		{
			if (decisionTree != null)
			{
				if (!decisionTree.IsForBoard(_board))
					throw new Exception("Decision tree for a different board");
				decisionTree.Reset();
			}
		}

		public void UndoLastDecision(Board _board)
		{
			if (decisionTree != null)
			{
				if (!decisionTree.IsForBoard(_board))
					throw new Exception("Decision tree for a different board");
				decisionTree.UndoLastDecision();
			}
		}

		protected int Solve(Board _board, int _maxIterations, bool _resetDecisionTree)
		{
			int iterations = 0;

			CheckConsistency(_board);

			SetIdentifiedValues(_board);
			if (_board.IsFinished())
				return iterations;

			if (_resetDecisionTree || decisionTree == null)
				decisionTree = new DecisionTree(_board);
			
			iterations = IterateSolution(_board, _maxIterations);

			CheckConsistency(_board);

			return iterations;
		}

		public int IterateSolution(Board _board, int _maxIterations)
		{
			int iterations = 0;
			while (_board.NotFinished())
			{
				if (iterations > _maxIterations)
				{
					throw new IterationExcess("Max number of iterations exceeded (" + iterations + ")");
				}
				iterations++;
				PerformNextDecision(_board);
			}
			return iterations;
		}

		public void PerformNextDecision(Board _board)
		{
			Decision decision = decisionTree.NextDecision();
			bool hasContraddiction = false;
			do
			{
				try
				{
					hasContraddiction = false;
					SetIdentifiedValues(_board, decision);
				}
				catch (Contraddiction contraddiction)
				{
					Console.Out.WriteLine(contraddiction.Message);
					decision = decisionTree.ChangeDecision();
					hasContraddiction = true;
				}
			}
			while (hasContraddiction);
		}

		public void SetIdentifiedValues(Board _board, Decision _decision)
		{
			bool performLoop;
			do 
			{
				performLoop = false;
				foreach (Cell cell in _board.Cells)
					performLoop = performLoop || cell.IdentifyAndSetValue(_decision);
			}
			while(performLoop);
		}

		public void SetIdentifiedValues(Board _board)
		{
			SetIdentifiedValues(_board, null);
		}

		public void CheckConsistency(Board _board)
		{
			foreach (Row_Column_Set row in _board.Rows)
			{
				CheckConsistencyInRowColumnSet(row, "row");
			}
			foreach (Row_Column_Set column in _board.Columns)
			{
				CheckConsistencyInRowColumnSet(column, "column");
			}
			foreach (Row_Column_Set squareSet in _board.SquareSets)
			{
				CheckConsistencyInRowColumnSet(squareSet, "squareSet");
			}
		}

		protected void CheckConsistencyInRowColumnSet(Row_Column_Set _rowColumnSet, string _type)
		{
			string[] oneToNine = new string[9];
			foreach (Cell cell in _rowColumnSet.Cells)
			{
				if (cell.IsValueSet())
				{
					int val = cell.Val;
					if (oneToNine[val-1] != null)
						throw new Inconsistency("Inconsistency in " + _type + " " + _rowColumnSet.Index);
					else
						oneToNine[val-1] = val.ToString();
				}
			}
		}
	}
}
